### PR TITLE
Make ``self-cls-assignment`` check tuple assignment

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Add a ChangeLog entry describing what your PR does.
- [x] If it's a new feature, or an important bug fix, add a What's New entry in
      `doc/whatsnew/<current release.rst>`.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description

Continuation of #5254 

Basically we only checked single assignments and will now also check tuple assignments.
Furthermore, I have removed some redundant code.

### CODE REVIEW

**Review by Pierre-Sassoulas — APPROVED** [review]
Nice cleanup on top of the fix ! 👏

File: pylint/checkers/variables.py [hunk 1]
```diff
@@ -2024,16 +2024,17 @@ def _store_type_annotation_names(self, node):
 
     def _check_self_cls_assign(self, node):
```
Comment (cdce8p) [inline pylint/checkers/variables.py:2025]:
```suggestion
    def _check_self_cls_assign(self, node: nodes.Assign) -> None:
```
While we are at it, it makes sense to also add type annotations.

File: pylint/checkers/variables.py [hunk 2]
```diff
@@ -2024,16 +2024,17 @@ def _store_type_annotation_names(self, node):
...
         """Check that self/cls don't get assigned"""
-        assign_names = {
-            target.name
-            for target in node.targets
-            if isinstance(target, nodes.AssignName)
-        }
+        assign_names = []
```
Comment (cdce8p) [inline pylint/checkers/variables.py:2027]:
```suggestion
        assign_names: Set[Union[str, None]] = set()
```
Why not use a `set` here?
Unfortunately `name` can still be `str | None`, although in practice it will always be `str`. Already deprecated in `astroid`.

File: pylint/checkers/variables.py [hunk 3]
```diff
@@ -2024,16 +2024,17 @@ def _store_type_annotation_names(self, node):
...
+        for target in node.targets:
+            if isinstance(target, nodes.AssignName):
+                assign_names.append(target.name)
```
Comment (cdce8p) [inline pylint/checkers/variables.py:2030]:
```suggestion
                assign_names.add(target.name)
```

File: pylint/checkers/variables.py [hunk 4]
```diff
@@ -2024,16 +2024,17 @@ def _store_type_annotation_names(self, node):
...
+            elif isinstance(target, nodes.Tuple):
+                assign_names.extend(
```
Comment (cdce8p) [inline pylint/checkers/variables.py:2032]:
```suggestion
                assign_names.update(
```

File: pylint/checkers/variables.py [hunk 5]
```diff
@@ -2024,16 +2024,17 @@ def _store_type_annotation_names(self, node):
...
+                    elt.name for elt in target.elts if isinstance(elt, nodes.AssignName)
+                )
         scope = node.scope()
         nonlocals_with_same_name = any(
-            child
-            for child in scope.body
-            if isinstance(child, nodes.Nonlocal) and assign_names & set(child.names)
+            child for child in scope.body if isinstance(child, nodes.Nonlocal)
```
Comment (cdce8p) [inline pylint/checkers/variables.py:2037]: It seems like we also need to do some more work on `node.scope()`.
```
Cannot access member "body" for type "LocalsDictNodeNG"
  Member "body" is unknown
```

File: pylint/checkers/variables.py [hunk 6]
```diff
@@ -2022,18 +2022,19 @@ def _store_type_annotation_names(self, node):
...
 
-    def _check_self_cls_assign(self, node):
+    def _check_self_cls_assign(self, node: nodes.Assign) -> None:
         """Check that self/cls don't get assigned"""
-        assign_names = {
-            target.name
-            for target in node.targets
-            if isinstance(target, nodes.AssignName)
-        }
+        assign_names: Set[Union[str, None]] = set()
```
Comment (cdce8p) [inline pylint/checkers/variables.py:2027]: `Set` and `Union` need to be added as imports.

**Review by cdce8p — APPROVED** [review]
Looks good! Can be merged once the imports are fixed.

File: pylint/checkers/variables.py (hunk 6 above)
Comment (DanielNoord) [reply pylint/checkers/variables.py]: Changed to use `Optional`. That should be fine as well right? I'll merge if you give the okay!

Comment (cdce8p) [reply pylint/checkers/variables.py]: Optional works too.
